crates/wasi-provider/src/lib.rs

```diff
@@ -205,45 +224,49 @@ impl WasiRuntime {
         args: Vec<String>,
         dirs: HashMap<String, Option<String>>,
         log_file_location: L,
-    ) -> Result<Self, failure::Error> {
-        debug!("before parse file: {:?}", std::env::current_dir().unwrap());
+    ) -> Result<(Self, BufReader<File>), failure::Error> {
```

Comment:
I feel uncomfortable with this change.

A constructor isn't generally intended to return a Self _wrapped in another thing_. In fact, that seems contrary to any constructor pattern I've seen. The idea of a constructor is that it creates _an instance of the object_, not an instance of some other thing with an instance of the object embedded somewhere in there.

In addition, returning a file handle seems awkward, since it commits you to having open file handles before the WASM itself has even been run. So you are initializing an object (that consumes system resources) before the system is asking you to commit to running this thing. This very well may paint us into a design corner in the future, where we have heavy objects hanging around early in the game.

When trying this same thing with wasm3, it appears that one can delay creating a file handle until `run()` is called, which means that we don't get a penalty imposed on us until the module is actually executed.

Reply:
Yeah, I wasn't sure myself either. The problem right now is that there is no way to get the handle once we kick things off because the `runtime` is moved onto the thread, so I couldn't call the `output` method anymore. I think the way to do this would be returning the file handle on run along with a future because right now `run` is a blocking call. I'll take another look tomorrow with fresh eyes to see what other ways there are to tackle this

Reply:
To get around the constructor issue, you can just call `let output = runtime.output()`, right? I left a comment inline to show you what I mean.